File: problems-adapt/0301-0400/0373_greatest-distance-to-land/solution_two_pass_dp.py
class Solution:
    def greatestDistanceToLand(self, grid: list[list[int]]) -> int:
        n = len(grid)
        # distance field: land starts at 0, water at a sentinel standing in
        # for infinity; any value above the largest possible distance
        # (2n - 2) is safe, and n * n is a convenient pick
        inf = n * n
        dist = [[inf] * n for _ in range(n)]
        has_land = False
        has_water = False
        for i in range(n):
            for j in range(n):
                if grid[i][j] == 1:
                    dist[i][j] = 0
                    has_land = True
                else:
                    has_water = True
        # all water (nothing to measure from) or all land (nothing to measure)
        if not has_land or not has_water:
            return -1
        # two-pass DP, first sweep: top-left to bottom-right, so every cell
        # relaxes against its up and left neighbors plus one step
        for i in range(n):
            for j in range(n):
                if i > 0 and dist[i - 1][j] + 1 < dist[i][j]:
                    dist[i][j] = dist[i - 1][j] + 1
                if j > 0 and dist[i][j - 1] + 1 < dist[i][j]:
                    dist[i][j] = dist[i][j - 1] + 1
        # second sweep: bottom-right to top-left, covering down and right;
        # the four directions together span every Manhattan path
        best = 0
        for i in range(n - 1, -1, -1):
            for j in range(n - 1, -1, -1):
                if i + 1 < n and dist[i + 1][j] + 1 < dist[i][j]:
                    dist[i][j] = dist[i + 1][j] + 1
                if j + 1 < n and dist[i][j + 1] + 1 < dist[i][j]:
                    dist[i][j] = dist[i][j + 1] + 1
                # land stays at 0, so a plain running max over the field works
                if dist[i][j] > best:
                    best = dist[i][j]
        return best

File: problems-originals/1101-1200/1162_as-far-from-land-as-possible/solution_two_pass_dp.go
func maxDistance(grid [][]int) int {
	n := len(grid)
	// distance field: land starts at 0, water at a sentinel standing in
	// for infinity; any value above the largest possible distance
	// (2n - 2) is safe, and n * n is a convenient pick
	inf := n * n
	dist := make([][]int, n)
	hasLand, hasWater := false, false
	for i := 0; i < n; i++ {
		dist[i] = make([]int, n)
		for j := 0; j < n; j++ {
			if grid[i][j] == 1 {
				dist[i][j] = 0
				hasLand = true
			} else {
				dist[i][j] = inf
				hasWater = true
			}
		}
	}
	// all water (nothing to measure from) or all land (nothing to measure)
	if !hasLand || !hasWater {
		return -1
	}
	// two-pass DP, first sweep: top-left to bottom-right, so every cell
	// relaxes against its up and left neighbors plus one step
	for i := 0; i < n; i++ {
		for j := 0; j < n; j++ {
			if i > 0 && dist[i-1][j]+1 < dist[i][j] {
				dist[i][j] = dist[i-1][j] + 1
			}
			if j > 0 && dist[i][j-1]+1 < dist[i][j] {
				dist[i][j] = dist[i][j-1] + 1
			}
		}
	}
	// second sweep: bottom-right to top-left, covering down and right; the
	// four directions together span every Manhattan path
	best := 0
	for i := n - 1; i >= 0; i-- {
		for j := n - 1; j >= 0; j-- {
			if i+1 < n && dist[i+1][j]+1 < dist[i][j] {
				dist[i][j] = dist[i+1][j] + 1
			}
			if j+1 < n && dist[i][j+1]+1 < dist[i][j] {
				dist[i][j] = dist[i][j+1] + 1
			}
			// land stays at 0, so a plain running max over the field works
			if dist[i][j] > best {
				best = dist[i][j]
			}
		}
	}
	return best
}
